### ISSUE TITLE
Inconsistency detected by ld.so when using PyQt5 with AA_ShareOpenGLContexts

### ISSUE THREAD

**Issue by ntninja** [flatpak/freedesktop-sdk-images#97]
Currently, when running any PyQt5-based app with the `AA_ShareOpenGLContexts` set within any `org.freedesktop.Runtime` based runtime environment, the following message will be displayed on app shutdown and the app appears to crash (but no data is lost):

```
Inconsistency detected by ld.so: dl-close.c: 811: _dl_close: Assertion `map->l_init_called' failed!
```

Running the same script used in the following example on my host system (Debian unstable) with the same PyQt5 & Qt5 binaries using virtualenv does cause the message above to be displayed:

*example.py*:

```py3
#!/usr/bin/python3
import os
import sys

class qt5:
	from PyQt5 import QtCore as core
	from PyQt5 import QtWidgets as widgets

qt5.core.QCoreApplication.setAttribute(qt5.core.Qt.AA_ShareOpenGLContexts)
qt5.widgets.QApplication(sys.argv)
```

*pyqt5-example.json*:

```json
{
	"app-id": "org.example.Example",
	"runtime": "org.freedesktop.Platform",
	"runtime-version": "1.6",
	"sdk": "org.freedesktop.Sdk",
	"command": "example.py",
	
	"finish-args": [
		 "--socket=x11",
		 "--socket=wayland"
	],
	
	"modules": [
		{
			"name": "electron-cash",
			"sources": [
				{
					"type": "file",
					"path": "example.py"
				}
			],
			"buildsystem": "simple",
			"build-commands": [
				"cp example.py /app/bin/"
			],
			"modules": [
				{
					"name": "pyqt5",
					"config-opts": [ "--disable-static", "--enable-x11" ],
					"no-autogen": true,
					"buildsystem": "simple",
					"build-commands": [ 
						"pip3 install --no-index --find-links \"file://${PWD}\" --prefix=/app PyQt5==5.9.2"
					],
					"sources": [
						{
							"type": "file",
							"url": "https://files.pythonhosted.org/packages/3a/c6/26270f5550f00920045c2f0b222a7d03d7a64382825c68bf0bb1a51d854c/PyQt5-5.9.2-5.9.3-cp35.cp36.cp37-abi3-manylinux1_x86_64.whl",
							"sha256": "5dab7244b6932606490b47a0692e03d7b72d89f7d3dfa3d147a8b34d6af72451"
						},
						{
							"type": "file",
							"url": "https://files.pythonhosted.org/packages/6b/46/2869369c8ec634ec9dc714b84834e9264acbe87eb7492328efccfa0c855e/sip-4.19.8-cp35-cp35m-manylinux1_x86_64.whl",
							"sha256": "e72955e12f4fccf27aa421be383453d697b8a44bde2cc26b08d876fd492d0174"
						}
					]
				}
			]
		}
	]
}
```

I also have a recipe to compile the bindings from source using the KDE5.9 SDK image available, if you prefer that instead.

This issue is likely related to: https://bugzilla.redhat.com/show_bug.cgi?id=1398370

**Comment by TingPing** [comment]
You are using giant binary blobs with who knows what ABI, it isn't supportable really and you should be building pyqt from source.

**Comment by ntninja** [comment]
I love you too. Quoting myself:

> I also have a recipe to compile the bindings from source using the KDE5.9 SDK image available, if you prefer that instead.

So here you go:

*example.py*:

```py3
#!/usr/bin/python3
import os
import sys

class qt5:
	from PyQt5 import QtCore as core
	from PyQt5 import QtWidgets as widgets

qt5.core.QCoreApplication.setAttribute(qt5.core.Qt.AA_ShareOpenGLContexts)
qt5.widgets.QApplication(sys.argv)
```

*pyqt5-example.json*:

```json
{
	"app-id": "org.example.Example",
	"runtime": "org.kde.Platform",
	"runtime-version": "5.9",
	"sdk": "org.kde.Sdk",
	"command": "example.py",
	
	"finish-args": [
		 "--socket=x11",
		 "--socket=wayland"
	],
	
	"modules": [
		{
			"name": "electron-cash",
			"sources": [
				{
					"type": "file",
					"path": "example.py"
				}
			],
			"buildsystem": "simple",
			"build-commands": [
				"cp example.py /app/bin/"
			],
			"modules": [
				{
					"name": "pyqt5",
					"config-opts": [ "--disable-static", "--enable-x11" ],
					"no-autogen": true,
					"buildsystem": "simple",
					"build-commands": [ 
						"python3 configure.py --assume-shared --confirm-license --no-designer-plugin --no-qml-plugin --sysroot=/app -w --sip=/app/bin/sip --sip-incdir=/app/include -verbose QMAKE_CFLAGS_RELEASE='-I/usr/include/python3.5m/' QMAKE_CXXFLAGS_RELEASE='-I/usr/include/python3.5m/'", 
						"make -j $(nproc)",
						"make install DESTDIR=/app"
					],
					"sources": [
						{
							"type": "archive",
							"url": "https://sourceforge.net/projects/pyqt/files/PyQt5/PyQt-5.10.1/PyQt5_gpl-5.10.1.tar.gz",
							"sha256": "9932e971e825ece4ea08f84ad95017837fa8f3f29c6b0496985fa1093661e9ef"
						}
					],
					"modules": [
						{
							"name": "sip",
							"sources": [
								{
									"type": "archive",
									"url": "https://sourceforge.net/projects/pyqt/files/sip/sip-4.19.8/sip-4.19.8.tar.gz",
									"sha256": "7eaf7a2ea7d4d38a56dd6d2506574464bddf7cf284c960801679942377c297bc"
								}
							],
							"buildsystem": "simple",
							"build-commands": [
								"python3 configure.py --bindir=/app/bin --destdir=/app/lib/python3.5/site-packages --incdir=/app/include --pyidir=/app/lib/python3.5/site-packages --sipdir=/app/share/sip",
								"make",
								"make install"
							],
							"cleanup": [
								"/bin",
								"/include"
							]
						}
					]
				}
			]
		}
	]
}
```

**Comment by ntninja** [comment]
Btw, I can reproduce the much simpler test case at: https://bugzilla.redhat.com/show_bug.cgi?id=1398370#c19
On my host system nothing happens, but the same binaries in a FreeDesktop 1.6 SDK container raise the assertion.

The patch at https://bugzilla.redhat.com/show_bug.cgi?id=1398370#c31 also looks like it should fix the issue but I have no idea how to test this or submit a PR for this. Can you maybe give me some guidance?

(It also appears this issue should have been at https://github.com/flatpak/freedesktop-sdk-base.)

**Comment by ntninja** [comment]
Fixed in the 18.08 runtime at https://gitlab.com/freedesktop-sdk/freedesktop-sdk.
